fs/smb/client/truenas_streams.c

```diff
@@ -0,0 +1,633 @@
+		return 0;
+	}
+
+	to_read = le64_to_cpu(info.EndOfFile);
+	if ((to_read >= XATTR_LARGE_SIZE_MAX) ||
+	    ((to_read + streams_samba_compat_enabled) > dstsz)) {
+		/*
+		 * The SMB protocol and MS-FSCC does not provide an upper-bound
+		 * on the maximum size of an alternate data stream. ReFS on
+		 * windows limits ADS max size to 64 KiB, MacOS has no
+		 * real limit on size of resource forks. ZFS doesn't
+		 * have an upper bound on xattr size (these get written as
+		 * files when they're too large), but since our APIs here
+		 * are somewhat terrible (no pwrite / pread equivalent) we
+		 * limit to XATTR_LARGE_SIZE_MAX to reduce maximum allocation
+		 * size that we need to perform.
+		 *
+		 * Typically an alternate data stream will be less than 1 KiB.
+		 */
+
+		SMB2_close(xid, tcon, fid.persistent_fid, fid.volatile_fid);
+		return -ERANGE;
+	}
+
+	rsize = tcon->ses->server->ops->negotiate_rsize(tcon,
+	    CIFS_SB(dentry->d_sb)->ctx);
+
+	for (total_read = 0; to_read > total_read; total_read += bytes_read) {
+		unsigned int len = min(to_read, rsize);
+		unsigned int retries = 0;
+		rc = -EAGAIN;
+		char *pbuf = dst + total_read;
+
+		io_parms = (struct cifs_io_parms) {
+			.persistent_fid = fid.persistent_fid,
+			.volatile_fid = fid.volatile_fid,
+			.pid = current->tgid,
+			.tcon = tcon,
+			.offset = total_read,
+			.length = len,
+		};
+
+		while ((rc == -EAGAIN) && ((retries += 1) < STREAM_MAX_RETRIES)) {
+			rc = SMB2_read(xid, &io_parms, &bytes_read, &pbuf,
+			    &buf_type);
+		}
+
+		// If we've read to end of stream, rc will be 0 and bytes_read 0
+		if (rc || (bytes_read == 0))
+			break;
+	}
+
+	if (streams_samba_compat_enabled) {
+		// Append terminating NULL to match Samba behavior
+		*(dst + total_read + 1) = '\0';
+	}
+
+	SMB2_close(xid, tcon, fid.persistent_fid, fid.volatile_fid);
+	if (rc == 0) {
+		if (streams_samba_compat_enabled) {
+			*pused = total_read + streams_samba_compat_enabled;
+			*(dst + total_read) = '\0';
+		} else {
+			*pused = total_read;
+		}
+	}
+
+	return rc;
+}
+
+static int
+get_stream_name(const char *name_in, char **name_out)
+{
+	char *stream_name, *suffix;
+
+	if (strcmp(name_in, DEFAULT_DATA_STREAM) == 0) {
+		return -EINVAL;
+	}
+
+	stream_name = kstrdup(name_in + STREAM_XATTR_LEN, GFP_KERNEL);
+	if (stream_name == NULL) {
+		return -ENOMEM;
+	}
+
+	suffix = strstr(stream_name, STREAM_SUFFIX);
+	if (suffix == NULL) {
+		// malformed stream name (no :DATA suffix)
+		kfree(stream_name);
+		return -EINVAL;
+	}
+
+	// chop off :DATA suffix before sending the open request
+	*suffix = '\0';
+	*name_out = stream_name;
+
+	return 0;
+}
+
+int set_stream_xattr(struct dentry *dentry, const char *full_path,
+		     unsigned int xid, struct cifs_tcon *tcon,
+		     const char *name, const void *value, size_t size,
+		     int flags)
+{
+	char *stream_name;
+	int rc = 0;
+
+	rc = get_stream_name(name, &stream_name);
+	if (rc) {
+		return rc;
+	}
+
+	if ((value == NULL) && (size == 0)) {
```

Comment:
I am not sure what is the semantics here, but checking both and with `&&` looks redundant and potentially unsafe here.  I would check one, and if needed, asserted another.